providers/implementations/digests/sha3_prov.c

```diff
@@ -249,6 +249,44 @@ static PROV_SHA3_METHOD kmac_s390x_md =
     } else {                                                                   \
         ctx->meth = sha3_generic_md;                                           \
     }
+#elif defined(__aarch64__)
+# include "arm_arch.h"
+
+static sha3_absorb_fn armsha3_sha3_absorb;
+
+size_t SHA3_absorb_cext(uint64_t A[5][5], const unsigned char *inp, size_t len,
+                    size_t r);
+/*-
+ * Hardware-assisted ARMv8.2 SHA3 extension version of the absorb()
+ */
+static size_t armsha3_sha3_absorb(void *vctx, const void *inp, size_t len)
+{
+    KECCAK1600_CTX *ctx = vctx;
+
+    return SHA3_absorb_cext(ctx->A, inp, len, ctx->block_size);
+}
+
+static PROV_SHA3_METHOD sha3_ARMSHA3_md =
+{
+    armsha3_sha3_absorb,
+    generic_sha3_final
+};
+/* Users can switch back to the generic code by clearing either of the bits */
+# define ARM_SHA3_CAPABLE                                                      \
+    ((OPENSSL_armcap_P & ARMV8_SHA3) &&                                        \
+     (OPENSSL_armcap_P & ARMV8_WORTH_USING_SHA3))
```

Comment:
This looks great - thank you for changing it

Given the two tests here, it seems that we could make things simpler by just having one test, which is what I'd thought `ARMV8_WORTH_USING_SHA3` could be, but you are right - "has SHA3" and "worth using" seem to be two different things. And neither of them really sound like "SHA3 capable" any more.

So (and sorry, I know it was my suggestion) would it be better if `ARMV8_WORTH_USING_SHA3` was called `ARMV8_HAVE_SHA3_AND_WORTH_USING`?

In `armcap.c` above that would be a rename only (since `ARMV8_WORTH_USING_SHA3` is only set if `ARMV8_SHA3` is set)

Then instead of having the macro `ARM_SHA3_CAPABLE` the `if` statements at lines 279 and 285 just do

```
# define SHA3_SET_MD(uname, typ)                                               \
    if (OPENSSL_armcap_P & ARMV8_HAVE_SHA3_AND_WORTH_USING) {                                                    \
        ctx->meth = sha3_ARMSHA3_md;                                           \
    } else {                                                                   \
        ctx->meth = sha3_generic_md;                                           \
    }
```

Reply:
> [users] can assign 'paradox' values. `ARMV8_HAVE_SHA3_AND_WORTH_USING=1` will override `ARMV8_SHA3=0`. Is this the expected behavior?

Yes, I would say so. If someone starts setting bits in the the environment variable then they get exactly what they ask for. The environment variable is a test / debug / fix mechanism, so we shouldn't try to second-guess what's wanted if it's set.

It also makes things easier for people testing on new hardware! They just set the bits to correspond to the code path they care about testing...

Reply:
> ARMV8_SHA3=0 seems to be a master switch considering its name...

It should really be called `ARMV8_FEAT_SHA3`, since it's set based on the presence of the CPU instruction (feature) but that's just the way all the feature bits have been named, so that would be a refactor for another day

Reply:
Thanks for explanation, updated :)